protocols/perf/src/protocol.rs

```diff
@@ -18,184 +18,184 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use async_trait::async_trait;
-use futures::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use futures_timer::Delay;
 use instant::Instant;
-use libp2p_request_response as request_response;
-use libp2p_swarm::StreamProtocol;
-use std::io;
-
-use crate::{RunDuration, RunParams};
-
-const BUF: [u8; 65536] = [0; 64 << 10];
-
-#[derive(Debug)]
-pub enum Response {
-    Sender(usize),
-    Receiver(RunDuration),
-}
-
-#[derive(Default)]
-pub struct Codec {
-    to_receive: Option<usize>,
-
-    write_start: Option<Instant>,
-    read_start: Option<Instant>,
-    read_done: Option<Instant>,
+use std::time::Duration;
+
+use futures::{
+    future::{select, Either},
+    AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, FutureExt, SinkExt, Stream, StreamExt,
+};
+
+use crate::{Finished, Progressed, Run, RunDuration, RunParams, RunUpdate};
+
+const BUF: [u8; 1024] = [0; 1024];
+const REPORT_INTERVAL: Duration = Duration::from_secs(1);
+
+pub(crate) fn send_receive<S: AsyncRead + AsyncWrite + Unpin + Send + 'static>(
+    params: RunParams,
+    stream: S,
+) -> impl Stream<Item = Result<RunUpdate, std::io::Error>> {
+    // Use a channel to simulate a generator. `send_receive_inner` can `yield` events through the
+    // channel.
+    let (sender, receiver) = futures::channel::mpsc::channel(0);
```

Comment:
I am a bit worried that a size of `0` might block the reading and writing loops, why not use 1?

Reply:
> The channel’s capacity is equal to buffer + num-senders. In other words, each sender gets a guaranteed slot in the channel capacity, and on top of that there are buffer “first come, first serve” slots available to all senders.

https://docs.rs/futures/latest/futures/channel/mpsc/fn.channel.html

A buffer of `0` with one sender implies a capacity of `1`. `1` is exactly the capacity we need here. Each time a `Progressed` event is put into the channel, we want it to be emitted to the outside. We don't want two `Progressed` events to be inside the channel buffer.

Does that make sense @thomaseizinger?

Reply:
> We don't want two `Progressed` events to be inside the channel buffer.

If for whatever reason the receiver doesn't get polled, you want to block writing on the stream before the new `Progressed` can be sent? Doesn't that mess with the performance measurements?

Reply:
> If for whatever reason the receiver doesn't get polled,

The receiver is held within the `send_receive`. Thus, if the receiver isn't polled, neither is the actual logic sending and receiving on the stream.